From ddcaf82d93e4f414355764fa2b3a8c634ec82b7e Mon Sep 17 00:00:00 2001
From: Rander Wang <rander.wang@linux.intel.com>
Date: Thu, 12 Jul 2018 15:58:46 +0800
Subject: [PATCH 1355/1600] ASoC: SOF: add Icelake support in SOF

Icelake shares the setting of cannonlake because no difference
between them in host

Signed-off-by: Rander Wang <rander.wang@linux.intel.com>
---
 sound/soc/sof/intel/Kconfig |  9 +++++++++
 sound/soc/sof/intel/hda.c   | 15 +++++++++++++++
 sound/soc/sof/sof-pci-dev.c | 20 ++++++++++++++++++++
 3 files changed, 44 insertions(+)

diff --git a/sound/soc/sof/intel/Kconfig b/sound/soc/sof/intel/Kconfig
index d9b4b6a..8575005 100644
--- a/sound/soc/sof/intel/Kconfig
+++ b/sound/soc/sof/intel/Kconfig
@@ -87,6 +87,15 @@ config SND_SOC_SOF_SKYLAKE
           Say Y if you have such a device.
           If unsure select "N".
 
+config SND_SOC_SOF_ICELAKE
+	tristate "SOF support for Icelake"
+	select SND_SOC_SOF_HDA_COMMON
+	help
+          This adds support for Sound Open Firmware for Intel(R) platforms
+          using the Icelake processors.
+          Say Y if you have such a device.
+          If unsure select "N".
+
 config SND_SOC_SOF_HDA_COMMON
 	tristate
 	select SND_SOC_SOF_PCI
diff --git a/sound/soc/sof/intel/hda.c b/sound/soc/sof/intel/hda.c
index 4f13431..a3a814a 100644
--- a/sound/soc/sof/intel/hda.c
+++ b/sound/soc/sof/intel/hda.c
@@ -292,6 +292,21 @@ static const struct sof_intel_dsp_desc chip_info[] = {
 	.ipc_ctl = CNL_DSP_REG_HIPCCTL,
 	.ops = &sof_cnl_ops,
 },
+{
+	/* Icelake */
+	.id = 0x34c8,
+	.cores_num = 4,
+	.cores_mask = HDA_DSP_CORE_MASK(0) |
+				HDA_DSP_CORE_MASK(1) |
+				HDA_DSP_CORE_MASK(2) |
+				HDA_DSP_CORE_MASK(3),
+	.ipc_req = CNL_DSP_REG_HIPCIDR,
+	.ipc_req_mask = CNL_DSP_REG_HIPCIDR_BUSY,
+	.ipc_ack = CNL_DSP_REG_HIPCIDA,
+	.ipc_ack_mask = CNL_DSP_REG_HIPCIDA_DONE,
+	.ipc_ctl = CNL_DSP_REG_HIPCCTL,
+	.ops = &sof_cnl_ops,
+},
 };
 
 static const struct sof_intel_dsp_desc *get_chip_info(int pci_id)
diff --git a/sound/soc/sof/sof-pci-dev.c b/sound/soc/sof/sof-pci-dev.c
index 152a3d3..5445d80 100644
--- a/sound/soc/sof/sof-pci-dev.c
+++ b/sound/soc/sof/sof-pci-dev.c
@@ -114,6 +114,19 @@ static struct sof_dev_desc kbl_desc = {
 };
 #endif
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_ICELAKE)
+static const struct sof_dev_desc icl_desc = {
+	.machines               = snd_soc_acpi_intel_icl_machines,
+	.resindex_lpe_base      = 0,
+	.resindex_pcicfg_base   = -1,
+	.resindex_imr_base      = -1,
+	.irqindex_host_ipc      = -1,
+	.resindex_dma_base      = -1,
+	.nocodec_fw_filename = "intel/sof-icl.ri",
+	.nocodec_tplg_filename = "intel/sof-icl-nocodec.tplg"
+};
+#endif
+
 static void sof_pci_fw_cb(const struct firmware *fw, void *context)
 {
 	struct sof_platform_priv *priv = context;
@@ -164,6 +177,9 @@ static const struct sof_ops_table mach_ops[] = {
 #if IS_ENABLED(CONFIG_SND_SOC_SOF_KABYLAKE)
 	{&kbl_desc, &sof_apl_ops},
 #endif
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_ICELAKE)
+	{&icl_desc, &sof_cnl_ops},
+#endif
 };
 
 static struct snd_sof_dsp_ops *sof_pci_get_ops(const struct sof_dev_desc *d)
@@ -337,6 +353,10 @@ static const struct pci_device_id sof_pci_ids[] = {
 	{ PCI_DEVICE(0x8086, 0x9d70),
 		.driver_data = (unsigned long)&skl_desc},
 #endif
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_ICELAKE)
+	{ PCI_DEVICE(0x8086, 0x34C8),
+		.driver_data = (unsigned long)&icl_desc},
+#endif
 	{ 0, }
 };
 MODULE_DEVICE_TABLE(pci, sof_pci_ids);
-- 
2.7.4

